### PR TITLE
Web Inspector: Add Function Breakpoints/Tracepoints (like Symbolic Breakpoints)

### PR DESCRIPTION
**Pull request by dcrousso**
#### 631ebd578babc70634907386b4f103a80f4c754f
<pre>
Web Inspector: Add Function Breakpoints/Tracepoints (like Symbolic Breakpoints)
<a href="https://bugs.webkit.org/show_bug.cgi?id=142914">https://bugs.webkit.org/show_bug.cgi?id=142914</a>
&lt;rdar://problem/20242025&gt;

Reviewed by Patrick Angle.

Existing JavaScript breakpoints are only useful so long as the developer knows the source code
location of the function. Web Inspector does provide a bunch of ways to find that location (e.g.
Search Tab, hovering a symbol when paused to see it&apos;s definition, etc.), but that makes adding a
breakpoint into a multi-step process. To make matters worse, it&apos;s not uncommon for there to be
different functions with the same name, meaning that if the developer wants to add breakpoints to
all of them it becomes a geometric problem (and that&apos;s not even considering the hassle of having to
duplicate any breakpoint configuration each time).

Symbolic breakpoints are a single-step way to add a single breakpoint that will be used across all
functions that share the same name, regardless of when they are created or where they are located.

* Source/JavaScriptCore/debugger/Debugger.h:
(JSC::Debugger::Observer::willEnter): Added.
(JSC::Debugger::Observer::applyBreakpoints): Added.
* Source/JavaScriptCore/debugger/Debugger.cpp:
(JSC::Debugger::forEachRegisteredCodeBlock): Added.
(JSC::Debugger::applyBreakpoints):
(JSC::Debugger::callEvent):
Add hooks to allow `InspectorDebuggerAgent` to also do things whenever a new `CodeBlock` is created
and when a new `CallFrame` is about to be entered (i.e. a function call).

* Source/JavaScriptCore/inspector/protocol/Debugger.json:
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.h:
(Inspector::InspectorDebuggerAgent::SymbolicBreakpoint::operator== const): Added.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::functionName): Added.
(Inspector::InspectorDebuggerAgent::addSymbolicBreakpoint): Added.
(Inspector::InspectorDebuggerAgent::removeSymbolicBreakpoint): Added.
(Inspector::InspectorDebuggerAgent::willEnter): Added.
(Inspector::InspectorDebuggerAgent::applyBreakpoints): Added.
(Inspector::InspectorDebuggerAgent::SymbolicBreakpoint::matches): Added.
Add commands for adding/removing symbolic breakpoints, leveraging the above `Debugger` hooks to mark
matching `CodeBlock` as having a (symbolic) breakpoint so that debug hooks will not be ignored.

* Source/WebInspectorUI/UserInterface/Controllers/DebuggerManager.js:
(WI.DebuggerManager):
(WI.DebuggerManager.prototype.async initializeTarget):
(WI.DebuggerManager.pauseReasonFromPayload):
(WI.DebuggerManager.prototype.symbolicBreakpointsForSymbol): Added.
(WI.DebuggerManager.prototype.addSymbolicBreakpoint): Added.
(WI.DebuggerManager.prototype.removeSymbolicBreakpoint): Added.
(WI.DebuggerManager.prototype._setSymbolicBreakpoint): Added.
(WI.DebuggerManager.prototype._removeSymbolicBreakpoint): Added.
(WI.DebuggerManager.prototype._handleSymbolicBreakpointDisabledStateChanged): Added.
(WI.DebuggerManager.prototype._handleSymbolicBreakpointEditablePropertyChanged): Added.
(WI.DebuggerManager.prototype._handleSymbolicBreakpointActionsChanged): Added.
Manage adding/removing symbolic breakpoints, including handling state changes for existing symbolic
breakpoints (e.g. disabled, actions, etc.).

* Source/WebInspectorUI/UserInterface/Models/SymbolicBreakpoint.js: Added.
(WI.SymbolicBreakpoint):
(WI.SymbolicBreakpoint.supported):
(WI.SymbolicBreakpoint.fromJSON):
(WI.SymbolicBreakpoint.prototype.get symbol):
(WI.SymbolicBreakpoint.prototype.get caseSensitive):
(WI.SymbolicBreakpoint.prototype.get isRegex):
(WI.SymbolicBreakpoint.prototype.get displayName):
(WI.SymbolicBreakpoint.prototype.get editable):
(WI.SymbolicBreakpoint.prototype.matches):
(WI.SymbolicBreakpoint.prototype.equals):
(WI.SymbolicBreakpoint.prototype.remove):
(WI.SymbolicBreakpoint.prototype.saveIdentityToCookie):
(WI.SymbolicBreakpoint.prototype.toJSON):
Create a model object to represent a symbolic breakpoint.

* Source/WebInspectorUI/UserInterface/Views/BreakpointPopover.js:
(WI.BreakpointPopover.show):
* Source/WebInspectorUI/UserInterface/Views/SymbolicBreakpointPopover.js: Added.
(WI.SymbolicBreakpointPopover):
(WI.SymbolicBreakpointPopover.get supportsEditing):
(WI.SymbolicBreakpointPopover.prototype.populateContent):
(WI.SymbolicBreakpointPopover.prototype.createBreakpoint):
(WI.SymbolicBreakpointPopover.prototype._updateSymbolCodeMirrorMode):
* Source/WebInspectorUI/UserInterface/Views/SymbolicBreakpointPopover.css: Added.
(.popover .edit-breakpoint-popover-content .symbol.editor):
(.popover .edit-breakpoint-popover-content .symbol.editor &gt; .CodeMirror):
(.popover .edit-breakpoint-popover-content .symbol.editor + label):
(.popover .edit-breakpoint-popover-content .symbol.editor + label + label):
* Source/WebInspectorUI/UserInterface/Views/SymbolicBreakpointTreeElement.js: Added.
(WI.SymbolicBreakpointTreeElement):
* Source/WebInspectorUI/UserInterface/Views/SymbolicBreakpointTreeElement.css: Added.
(.item.breakpoint.symbolic:not(.paused) .icon):
(@media(prefers-color-scheme: dark) .item.breakpoint.symbolic:not(.paused) .icon):
* Source/WebInspectorUI/UserInterface/Views/NavigationSidebarPanel.js:
(WI.NavigationSidebarPanel.prototype._isTreeElementWithoutRepresentedObject):
* Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js:
(WI.SourcesNavigationSidebarPanel):
(WI.SourcesNavigationSidebarPanel.prototype.closed):
(WI.SourcesNavigationSidebarPanel.prototype.willDismissPopover):
(WI.SourcesNavigationSidebarPanel.prototype._willDismissSymbolicBreakpointPopover): Added.
(WI.SourcesNavigationSidebarPanel.prototype._insertDebuggerTreeElement):
(WI.SourcesNavigationSidebarPanel.prototype._addBreakpoint):
(WI.SourcesNavigationSidebarPanel.prototype._updatePauseReasonSection):
(WI.SourcesNavigationSidebarPanel.prototype._handleTreeSelectionDidChange):
(WI.SourcesNavigationSidebarPanel.prototype._populateCreateBreakpointContextMenu):
Add various UI elements for `WI.SymbolicBreakpoint`.

* Source/WebInspectorUI/UserInterface/Base/Setting.js:
* Source/WebInspectorUI/UserInterface/Views/SettingsTabContentView.js:
(WI.SettingsTabContentView.prototype._createExperimentalSettingsView):
Create an experimental setting for symbolic breakpoints, as they currently do not work for
- native functions (e.g. `Event.prototype.preventDefault`)
- functions that changed their `name` after creation (e.g. `foo.name = &quot;bar&quot;`)

* Source/WebInspectorUI/UserInterface/Base/ObjectStore.js:
(WI.ObjectStore._open):
Increment the database version for the new symbolic breakpoint object store.

* Source/WebInspectorUI/UserInterface/Base/ReferencePage.js:
Create a link to a reference page (to be created).

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Images/TypeIcons.svg:
* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:

* Source/WebInspectorUI/UserInterface/Test.html:
* LayoutTests/inspector/debugger/symbolic-breakpoint.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-exact-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-exact-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-exact-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-exact-case-sensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-regex-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-regex-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-regex-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-regex-case-sensitive-expected.txt: Added.

Canonical link: <a href="https://commits.webkit.org/253374@main">https://commits.webkit.org/253374@main</a>
</pre>

### CODE REVIEW

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp [hunk 1]
```diff
@@ -661,6 +661,78 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
...
+Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeSymbolicBreakpoint(const String& symbol, std::optional<bool>&& caseSensitive, std::optional<bool>&& isRegex)
+{
+    SymbolicBreakpoint symbolicBreakpoint;
+    symbolicBreakpoint.symbol = symbol;
+    if (caseSensitive)
+        symbolicBreakpoint.caseSensitive = *caseSensitive;
+    if (isRegex)
+        symbolicBreakpoint.isRegex = *isRegex;
+
+    if (!m_symbolicBreakpoints.removeAll(symbolicBreakpoint))
```
Comment (JosephPecoraro) [inline Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:733]: Is the `removeAll` just being safe? It seems we already error if there were duplicates.

Comment (dcrousso) [reply Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:733]: yeah it's just being safe.  we do this for blackbox configs too, so i figured i'd also match it :)

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp [hunk 2]
```diff
@@ -1460,4 +1576,14 @@ void InspectorDebuggerAgent::clearAsyncStackTraceData()
...
 }
 
+bool InspectorDebuggerAgent::SymbolicBreakpoint::matches(const String& symbol) const
+{
+    if (symbol.isEmpty())
+        return false;
+
+    auto searchStringType = isRegex ? ContentSearchUtilities::SearchStringType::Regex : ContentSearchUtilities::SearchStringType::ExactString;
+    auto regex = ContentSearchUtilities::createRegularExpressionForSearchString(this->symbol, caseSensitive, searchStringType);
+    return regex.match(symbol) != -1;
```
Comment (saambarati) [inline Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp]: nit: Should we just cache the RegularExpression on the SymbolicBreakpoint struct? We might be calling this a lot

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp [hunk 3]
```diff
@@ -661,6 +661,78 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
...
+    if (!codeBlock)
+        return nullString();
+
+    if (codeBlock->codeType() != JSC::CodeType::FunctionCode)
+        return nullString();
+
+    return JSC::jsCast<JSC::FunctionExecutable*>(codeBlock->ownerExecutable())->ecmaName().string();
+}
+
+Protocol::ErrorStringOr<void> InspectorDebuggerAgent::addSymbolicBreakpoint(const String& symbol, std::optional<bool>&& caseSensitive, std::optional<bool>&& isRegex, RefPtr<JSON::Object>&& options)
```
Comment (saambarati) [inline Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:686]: Why are these optionals passed in as rvalues?

Comment (dcrousso) [reply Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:686]: that's just how the code generated by the Web Inspector protocol JSON does it 😅

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp [hunk 4]
```diff
@@ -661,6 +661,78 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
...
+
+    auto& symbolicBreakpoint = m_symbolicBreakpoints.last();
+
+    m_debugger.applyToAllRegisteredCodeBlocks([&] (JSC::CodeBlock* codeBlock) {
+        auto name = getFunctionName(codeBlock);
+
+        if (symbolicBreakpoint.matches(name)) {
+            symbolicBreakpoint.knownMatchingSymbols.add(name);
+
+            JSC::JSLockHolder locker(m_debugger.vm());
```
Comment (saambarati) [inline Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp]: I believe you need to be holding the JS lock to iterate all code blocks.

**Comment by dcrousso** [comment]
Screenshot 2022-08-08 at 23 31 03

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp [hunk 5]
```diff
@@ -661,6 +661,90 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
...
+        symbolicBreakpoint.specialBreakpoint = WTFMove(breakpoint);
+
+        if (!m_symbolicBreakpoints.appendIfNotContains(WTFMove(symbolicBreakpoint)))
+            return makeUnexpected("Symbolic breakpoint for given symbol, given caseSensitive, and given isRegex already exists"_s);
+    }
+
+    auto& symbolicBreakpoint = m_symbolicBreakpoints.last();
+
+    {
+        JSC::JSLockHolder locker(m_debugger.vm());
```
Comment (patrickangle) [inline Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:710]: It appears we must hold this lock to iterate code blocks, so can we just take the lock inside `forEachRegisteredCodeBlock` instead? Unless you think we will at some point start using the function in code that already holds the lock...

Aside: It's unfortunate there isn't currently a way (that I can find) to compile-time or runtime enforce a `JSLock`.

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp [hunk 6]
```diff
@@ -661,6 +661,90 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
...
+
+    if (!m_symbolicBreakpoints.removeAll(symbolicBreakpoint))
+        return makeUnexpected("Missing symbolic breakpoint for given symbol, given caseSensitive, and given isRegex"_s);
+
+    {
+        JSC::JSLockHolder locker(m_debugger.vm());
+
+        m_debugger.forEachRegisteredCodeBlock([&] (JSC::CodeBlock* codeBlock) {
+            if (symbolicBreakpoint.matches(functionName(codeBlock)))
+                codeBlock->removeBreakpoint(1);
```
Comment (patrickangle) [inline Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:741]: Does this work if a named function is created after the breakpoint was created but before it was removed? Can we add a test for that sequence of events just to make sure? e.g.
1. Create breakpoint for `foo`
2. Create a new class with a function named `foo`
3. Create an instance of that class and call `foo` to make sure we break
4. Remove the breakpoint and make sure the world doesn't burn

File: Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js [hunk 7]
```diff
@@ -851,6 +855,11 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
             return;
         }
 
+        if (popover instanceof WI.SymbolicBreakpointPopover) {
+            this._willDismissSymbolicBreakpointPopover(popover);
+            return;
+        }
+
         console.assert();
```
Comment (patrickangle) [inline Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js]: While we are here, can we add the popover to the output of this assertion? 😅

File: Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js [hunk 8]
```diff
@@ -1794,6 +1818,31 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
...
+            let symbolicBreakpoint = WI.debuggerManager.symbolicBreakpointsForSymbol(pauseData.name)[0];
+            console.assert(symbolicBreakpoint, "Expected Symbolic breakpoint for function name.", pauseData.name);
+            if (!symbolicBreakpoint)
+                break;
+
+            this._pauseReasonTreeOutline = this.createContentTreeOutline({suppressFiltering: true});
+
+            let symbolicBreakpointTreeElement = new WI.SymbolicBreakpointTreeElement(symbolicBreakpoint, {
+                classNames: ["paused"],
+                title: WI.UIString("Calling Function \u201C%s\u201D").format(pauseData.name),
```
Comment (patrickangle) [inline Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js]: A localization comment would be nice with this (and maybe a unique key as well)

File: Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js [hunk 9]
```diff
@@ -2041,6 +2091,15 @@ WI.SourcesNavigationSidebarPanel = class SourcesNavigationSidebarPanel extends W
             }, !!breakpoint);
         }
 
+        if (WI.SymbolicBreakpoint.supported()) {
+            contextMenu.appendItem(WI.UIString("Symbolic Breakpoint\u2026"), () => {
```
Comment (patrickangle) [inline Source/WebInspectorUI/UserInterface/Views/SourcesNavigationSidebarPanel.js]: Ditto :1835

**Review by patrickangle — APPROVED** [review]
Awesome stuff. I agree with making this experimental for now mainly because you can't set a symbolic breakpoint on native functions. Functions where `name` is set later is less concerning to me because the function in the source code we have will still have the same `ecmaName` that you are using in this patch.

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp (hunk 5 above)
Comment (dcrousso) [reply Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:710]: according to @saambarati, we should also be holding the lock when calling `forEachRegisteredCodeBlock` too

File: Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp (hunk 6 above)
Comment (dcrousso) [reply Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:741]: we actually already test that :)

`CodeBlock` are created lazily, so the `triggerSymbolicBreakpoint` won't actually attempt to create it until the first time it's invoked, which happens after we `addSymbolicBreakpoint` in the test.  i.e. the first test case will cause the creation of the `CodeBlock` and the rest will reuse it

**Pull request merged**